Assets/MixedRealityToolkit/Utilities/Rendering/MaterialInstance.cs

```diff
@@ -0,0 +1,305 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+#if UNITY_EDITOR
+using UnityEditor;
+#endif
+
+namespace Microsoft.MixedReality.Toolkit.Rendering
+{
+    /// <summary>
+    /// The MaterialInstance behavior aides in tracking instance material lifetime and automatically destroys instanced materials for the user. 
+    /// This utility component can be used as a replacement to <see href="https://docs.unity3d.com/ScriptReference/Renderer-material.html">Renderer.material</see> or 
+    /// <see href="https://docs.unity3d.com/ScriptReference/Renderer-materials.html">Renderer.materials</see>. When invoking Unity's Renderer.material(s), Unity 
+    /// automatically instantiates new materials. It is the caller's responsibility to destroy the materials when a material is no longer needed or the game object is 
+    /// destroyed. The MaterialInstance behavior helps avoid material leaks and keeps material allocation paths consistent during edit and run time.
+    /// </summary>
+    [HelpURL("https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Rendering/MaterialInstance.html")]
+    [ExecuteAlways, RequireComponent(typeof(Renderer))]
+    public class MaterialInstance : MonoBehaviour
```

Comment:
Looks like this may needs to be consolidated with UIMaterialInstantiator from PR #6049

Reply:
Good idea, it doesn't look like UIMaterialInstantiator ever destroys the material either? Once this PR is in there are a few places I want to remove instances of Renderer.Material(s). I'll enter a bug for this to track it.

Reply:
Opened an issue: https://github.com/microsoft/MixedRealityToolkit-Unity/issues/6166